File: 141-linked-list-cycle/141-linked-list-cycle.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def hasCycle(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        slow = fast = head # Slow and fast start at beginning of list.
        
        while slow and fast and fast.next:
            
            slow = slow.next
            fast = fast.next.next
            
            if slow == fast: # Fast caught up with slow so cycle!
                return True
            
        return False # Fast didn't catch slow before either reached None.
        
        
        
        
        

File: 141-linked-list-cycle/141-linked-list-cycle.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    
    public boolean hasCycle(ListNode head) {
        
        ListNode slow = head, fast = head; // Slow and fast start at beginning of list.
        
        while (slow != null && fast != null && fast.next != null) {
            
            slow = slow.next;
            fast = fast.next.next;
            if (slow == fast) return true; // Fast caught up with slow so cycle!   
                
        }
            
        return false; // Fast didn't catch slow before either reached None.
        
    }
    
}